src/uct/ib/rdmacm/rdmacm_cm_ep.c

```diff
@@ -157,97 +157,70 @@ ucs_status_t uct_rdmacm_cm_ep_conn_notify(uct_ep_h ep)
-    if (ucs_list_is_empty(&ctx->reserved_qpn_ctx.blk_list) ||
+    blk = ucs_list_tail(&ctx->blk_list, struct uct_rdmacm_cm_reserved_qpn_blk, entry);
+    if (ucs_list_is_empty(&ctx->blk_list) ||
         (blk->next_avail_qpn_offset >= qpns_per_obj)) {
-        status = uct_rdmacm_cm_reserved_qpn_blk_alloc(uct_rdmacm_cm_ep_get_cm(cep),
-                                                      cep->id->verbs, ctx, &blk);
+        status = uct_rdmacm_cm_reserved_qpn_blk_add(cep->id->verbs, ctx, &blk);
         if (status != UCS_OK) {
             goto out;
         }
 
-        ucs_list_add_tail(&ctx->reserved_qpn_ctx.blk_list, &blk->entry);
+        ucs_list_add_tail(&ctx->blk_list, &blk->entry);
     }
 
     cep->qpn = blk->first_qpn + blk->next_avail_qpn_offset;
     ucs_debug("created reserved qp_num: 0x%x on rdmacm_id %p",
               cep->qpn, cep->id);
-    UCS_BITMAP_SET(blk->bmp, blk->next_avail_qpn_offset);
+    cep->blk = blk;
+    blk->refcount++;
     blk->next_avail_qpn_offset++;
+    status = UCS_OK;
 
 out:
-    pthread_mutex_unlock(&ctx->reserved_qpn_ctx.mutex);
+    ucs_spin_unlock(&ctx->lock);
     return status;
 }
 
 /**
  * Deallocate a reserved QPN. The FW object is destroyed only when all QPNs
  * in this object were used and freed.
  */
-static void uct_rdamcm_cm_ep_destroy_reserved_qpn(uct_rdmacm_cm_ep_t *cep)
+static void
+uct_rdamcm_cm_ep_destroy_reserved_qpn(uct_rdmacm_cm_ep_t *cep,
+                                      struct uct_rdmacm_cm_device_context *ctx)
 {
-    struct uct_rdmacm_cm_qp_context *ctx;
-    struct uct_rdmacm_cm_reserved_qpn_blk *blk;
     uint32_t qpns_per_obj;
-    bool found;
     ucs_status_t status;
 
+    /* Reserved QP number is created on-demand when we send the private data */
     if (!(cep->flags & UCT_RDMACM_CM_EP_QPN_CREATED)) {
         return;
     }
 
-    ucs_assert_always(cep->qpn != 0);
-
     ucs_debug("cm ep destroy reserved qpn on rdmacm_id %p", cep->id);
 
-    status = uct_rdmacm_cm_get_qp_context(uct_rdmacm_cm_ep_get_cm(cep),
-                                          cep->id->verbs, &ctx);
-    if (status != UCS_OK) {
-        ucs_error("cm ep destroy reserved qpn failed on rdmacm_id %p: failed get qp context",
-                  cep->id);
-        return;
-    }
-
-    qpns_per_obj = uct_rdmacm_cm_get_qpns_per_obj(ctx->reserved_qpn_ctx);
+    qpns_per_obj = UCS_BIT(ctx->log_reserved_qpn_granularity);
 
-    pthread_mutex_lock(&ctx->reserved_qpn_ctx.mutex);
+    ucs_spin_lock(&ctx->lock);
 
-    found = false;
-    ucs_list_for_each(blk, &ctx->reserved_qpn_ctx.blk_list, entry) {
-        if ((cep->qpn >= blk->first_qpn) &&
-            (cep->qpn < blk->first_qpn + qpns_per_obj)) {
-            found = true;
-            break;
+    --cep->blk->refcount;
+    if ((cep->blk->next_avail_qpn_offset >= qpns_per_obj) &&
+        (cep->blk->refcount == 0)) {
+        status = uct_rdmacm_cm_reserved_qpn_blk_destroy(cep->blk);
```

Comment:
uct_rdmacm_cm_reserved_qpn_blk_destroy should return void , and move the warning printing inside it